[bobscodedump] iP

bobscodedump opened the pull request:
# Updated iP PR Description 🖥️

## Description

This pull request (PR) aims to implement several enhancements and bug fixes for the iP project. Below is a summary of the changes and tasks in this PR:

### Changes in this PR:

- Added a new feature to the iP project to improve user experience.
- Fixed a critical bug that caused unexpected behavior.
- Updated the project documentation for clarity and completeness.

### Tasks to be completed:

1. [x] Implement the new feature.
2. [ ] Added Validation
3. [ ] Update the documentation with the latest changes.

Main Method

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

bobscodedump committed:
Level-0

bobscodedump committed:
Level-1

bobscodedump committed:
Level-2

bobscodedump committed:
Level-3

bobscodedump committed:
Level-4

bobscodedump committed:
Level-5

bobscodedump committed:
Level-6

bobscodedump committed:
A-TextUiTesting

Commit:
Level-7

Commit:
Level-7

Commit:
Level-7

Commit:
Revert "Level-7"

This reverts commit db6f80b6012d65c715e006bdd170e3e51d4637fa.

Commit:
branch-level-7

Commit:
merge to master

Commit:
Level-8

Commit:
fixed bug for Level-8

Commit:
Merge branch 'add-gradle-support'

merge

Commit:
A-Gradle

Commit:
A-JUnit

Commit:
Level-9

Commit:
A-Checkstyle

dom-buri commented on src/test/java/duke/DateParserServiceTest.java:
```diff
@@ -0,0 +1,41 @@
+
+    @Test
+    public void testValidDateFormats() {
+        // Test various valid date formats
```
Should the comment start with "Tests" instead?

dom-buri commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,116 @@
+     * @throws DukeException If there is an error while loading tasks.
+     */
+    public ArrayList<Task> load() throws DukeException {
+        ArrayList<Task> dukeList = new ArrayList<>();
+        String[] splitFilePath = this.filePath.split("/");
+        String foldersPath = String.join("/", Arrays.copyOf(splitFilePath, splitFilePath.length - 1));
+        new File(foldersPath).mkdirs();
+        boolean fileExists;
```
I like how you name your variable name. They are all in camelCase and easy to understand. The boolean fileExists also sound like a boolean.

dom-buri commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,127 @@
+     * @param size The updated size of the task list.
+     */
+    public void addToList(Task task, int size) {
+        System.out.println("Got it. I've added this task:\n\t" + task.toString()
+                + "\nNow you have " + size + " tasks in the list.");
```
I like how you placed line break correctly to improve readability.

dom-buri commented on src/test/java/duke/ParserTest.java:
```diff
@@ -0,0 +1,63 @@
+public class ParserTest {
+
+    @Test
+    public void testParse_validCommands() {
```
Should the method names be testParseValidCommands() instead similar to DateParserServiceTest.java instead?

peasantbird replied:
Perhaps the method name for testParse_ValidCommands() should follow the original method name in Parser.java?

dom-buri reviewed (commented):
LGTM. I found your code easy to read and understand. The code also followed the Java coding standard very closely. However, I noted a few minor coding standard violations in the Javadoc as well as in the JUnit test files.

peasantbird commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,116 @@
+            String line;
+            try {
+                while ((line = reader.readLine()) != null) {
+                    if (line.startsWith("T")) {
+                        String[] inputs = line.split(" \\| ", 3);
+                        boolean isDone = inputs[1].equals("1");
+                        ToDo todo = new ToDo(inputs[2], isDone);
+                        dukeList.add(todo);
+                    }
+                    if (line.startsWith("D")) {
+                        String[] inputs = line.split(" \\| ", 4);
+                        boolean isDone = inputs[1].equals("1");
+                        LocalDate b;
+                        try {
+                            b = DateParserService.parseDate(inputs[3]);
+                        } catch (DateTimeParseException e) {
+                            throw new DukeException("Invalid date format");
+                        }
+                        Deadline deadline = new Deadline(inputs[2], isDone, b);
+                        dukeList.add(deadline);
+                    }
+                    if (line.startsWith("E")) {
+                        String[] inputs = line.split(" \\| ", 5);
+                        boolean isDone = inputs[1].equals("1");
+                        LocalDate f;
+                        LocalDate t;
+                        try {
+                            f = DateParserService.parseDate(inputs[3]);
+                            t = DateParserService.parseDate(inputs[4]);
+                        } catch (DateTimeParseException e) {
+                            throw new DukeException("Invalid date format");
+                        }
+                        Event event = new Event(inputs[2], isDone, f, t);
+                        dukeList.add(event);
+                    }
```
Not sure I like the long load() method. Perhaps the logic for parsing each type of task from the save file could be placed in another function(s)?

peasantbird commented on src/main/java/duke/DukeList.java:
```diff
@@ -0,0 +1,81 @@
+     *
+     * @return A formatted string containing the tasks in the list.
+     */
+    @Override
+    public String toString() {
+        StringBuilder str = new StringBuilder();
+        for (int i = 1; i <= dukeList.size(); i++) {
+            Task task = dukeList.get(i - 1);
+            str.append(i).append(". ").append(task.toString()).append("\n");
+        }
+        return str.toString();
+    }
```
I like how you extracted out the logic of printing the task list into a toString() method!

peasantbird commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,57 @@
+     * @throws DukeException If the user command is not recognized or has incorrect format.
+     */
+    public static Command parse(String command) throws DukeException {
+        if (command.equals("list")) {
+            return new DisplayCommand(command);
+        }
+        if (command.startsWith("mark")) {
+            return new MarkCommand(command);
+        }
+        if (command.startsWith("unmark")) {
+            return new UnMarkCommand(command);
+        }
+        if (command.startsWith("todo")) {
+            return new AddToDoCommand(command);
+        }
+        if (command.startsWith("deadline")) {
+            return new AddDeadlineCommand(command);
+        }
+        if (command.startsWith("event")) {
+            return new AddEventCommand(command);
+        }
+        if (command.startsWith("delete")) {
+            return new DeleteCommand(command);
+        }
+        if (command.startsWith("find")) {
+            return new FindCommand(command);
+        }
+        if (command.equals("bye")) {
+            return new ByeCommand(command);
+        }
+        throw new DukeException("Unrecognized command :(");
+    }
```
Perhaps a switch-case would be syntactically nicer?

peasantbird reviewed (commented):
Overall, I found your code easy to read because you separated the closely-related code into their classes. There were really only a few places where the nesting was too deep, and just me nitpicking a little... LGTM!

peasantbird commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,127 @@
+    /**
+     * Displays a message indicating that the list of matching tasks is about to be shown.
+     */
+    public void matchingTasks() {
```
Perhaps a verb should be used to name this function?

peasantbird commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,90 @@
+     *
+     * @return A formatted string containing task details for file storage.
+     */
+    public String writeFile() {
```
Perhaps a better method name, because the file is not being written here?

peasantbird commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,127 @@
+     * @param task The task that was added.
+     * @param size The updated size of the task list.
+     */
+    public void addToList(Task task, int size) {
```
Perhaps a better method name, because a task is not being added here?

peasantbird commented on src/main/java/duke/command/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,57 @@
+            deadline = new Deadline(by[0], b);
+        } catch (ArrayIndexOutOfBoundsException | DateTimeParseException e) {
+            if (e instanceof ArrayIndexOutOfBoundsException) {
+                throw new DukeException("Invalid Deadline format");
+            }
+            throw new DukeException("Invalid date format");
+        }
+        tasks.add(deadline);
+        try {
+            storage.writeToFile(tasks);
+        } catch (IOException e) {
+            throw new DukeException("Error writing to file");
```
Perhaps you could also make some subclasses of DukeException to better describe these exceptions that occur in your chatbot!

peasantbird reviewed (commented):
LGTM! Just a bit of nitpicking of some namings. Apart from that, OOP was implemented really well, making your code easy to read and understand. On top of that, your Javadocs is really well-written and clear, and your JUnit tests are very thorough!

Commit:
Level-10

Commit:
Added assertions

Added assertions to various methods to ensure important assumptions are upheld, and make sure errors are pointed out

Commit:
A-CodingQuality

Commit:
merge

Commit:
A-CI

Commit:
C-Tagging

Commit:
fixed code styling

Commit:
changed workflow

Commit:
made some final changes

changes to code quality, and added a welcome message

Commit:
A-UserGuide

Commit:
added ui screenshot

Commit:
changed ui screenshot

bobscodedump committed:
Update README.md

bobscodedump committed:
A-Release

bobscodedump committed:
merge
